### ISSUE TITLE
MIDI Recorder consuming 100% CPU

### ISSUE THREAD

**Issue by riban-bw** [zynthian/zynthian-issue-tracking#535]
**Describe the bug**
Any access to MIDI recorder seems to run the zynthian-ui CPU usage up to 100% for a long time. This includes accessing the MIDI Recorder page or toggling MIDI record via CUIA.

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to MIDI Recorder page

**Expected behaviour**
MIDI Recorder page shows.

**Actual behaviour**
Zynthian freezes for many 10's of seconds before MIDI Recorder page shows.

**Setup:**
- HARDWARE
Raspberry Pi 3 Model B Plus Rev 1.3
Soundcard: HifiBerry DAC+ ADC PRO
Display: ZynScreen 3.5 (v1)
Wiring: MCP23017_ZynScreen
GPIO Expander: MCP23017
- SYSTEM
Raspbian GNU/Linux 10 (buster)
Build Date: 2021-09-23
Memory: 41% (385M/923M)
SD Card: 40% (11G/29G)
Temperature: 56.9ºC
- MIDI
Profile: default.sh
Fine Tuning: 440 Hz
Master Channel: 16
Single Active Channel: off
ZS3 SubSnapShots: on
- SOFTWARE
zyncoder: testing (7c09738)
zynthian-ui: feature/516_launchkey (37c633a)
zynthian-sys: testing (b49f8e8)
zynthian-data: testing (006a071)
zynthian-webconf: testing (39b61bb)

**Additional context**
I observed this whilst toggling MIDI record from LaunchKey Mini Mk3. I don't recall this occurring during previous work on this branch. It may relate to moving development to Raspberry Pi 3, the content of the capture directory, or something else. Capture directory:

```
total 5240
-rw-r--r-- 1 root root  525334 Oct  6 17:43 '001-JV;Calf Monosynth;OctaWorm.ogg'
-rw-r--r-- 1 root root 3479640 Oct  9 11:58 '001-JV;Calf Monosynth;OctaWorm.wav'
-rw-r--r-- 1 root root 1198168 Oct  9 12:01 '002-JV;Calf Monosynth;OctaWorm.wav'
-rw-r--r-- 1 root root   67363 Sep  5  2020  Bladerunner-1.mid
-rw-r--r-- 1 root root   36844 Sep 23 00:20 'House In RTP.mid'
-rw-r--r-- 1 root root   45201 Sep  5  2020  roland_take5.mid
lrwxrwxrwx 1 root root      16 Sep  5  2020  test.mid -> roland_take5.mid
```

**Comment by riban-bw** [comment]
Each time a SMF file is read it consumes CPU whilst reading the content. Each time a SMF is played, stopped, etc. the zyngui screen is updated for all SMF files which reads them all. Need to figure out what is consuming CPU whilst reading file and reduce the quantity of reads required to refresh screen. This is very evident on a Raspberry Pi 3 with a few SMF each a few minutes long.

**Comment by riban-bw** [comment]
When a SMF file is opened, each event is inserted into a schedule formed of a std::vector. There are two expensive actions:

1. The vector is iterated from start until insert time is detected which probably means the vector gets iterated from start to end for every event added.
2. Events are inserted into the vector which necessitates creating a new vector and copying due to the nature of a vector.

This occurs in Track::addEvent method.

**Comment by riban-bw** [comment]
I tried changing from use of std::vector to std::list with no discernible improvement. I changed the insert to search from end of sequence backwards which made a substantial improvement. It may still be advantageous to avoid the expensive resizing of vector by using a linked list but for now the main issue is fixed in branch _feature/516_launchkey_. If necessary this can be split to its own fix branch and applied to testing. It may be required as a hot-fix for stable. For now I will leave it in the launchkey branch pending request from project lead.

**Comment by jofemodo** [comment]
I can confirm this bug is present when using RBPi4 and i don't recall having this issue previously on previous stable or testing. I would concentrate on latest changes, or perhaps some library used by the SMF player.

**Comment by riban-bw** [comment]
I wonder if there is a change in the kernel / GCC / standard libraries because this is the very reason we changed to libzynsmf (slow reading of SMF) and the code worked fine. Something changed which I believe to be in underlying OS / libs because I don't think any work was done on SMF. I checked recent history and could not see any changes. The fix does optimise and makes the perceived problem go away so should be there anyway but you are right, something odd has happened here.

**Comment by riban-bw** [comment]
I have implemented the reverse insert in branch bug/535_smf_slow_file_read. I need to validate this is working (I did this in the airport without a real Zynthian to hand) and if okay will submit a pull request.

**Comment by jofemodo** [comment]
OK! Tell me when you want i test it ;-)

**Comment by riban-bw** [comment]
Okay - the fix works but exposes other issues including:
- Tempo changes not being implemented properly (#543)
- Duplicating the list of recordings until the screen is left then re-entered (#550)
- Not resetting controllers, e.g. pitch bend (#549)
- Playback highlight jumps to top (#551)

These are all reported separately so I suggest testing this fix and merging to testing if satisfactory but maybe wait until some of the others are resolved before producing a hot-fix for stable (which is required soon due to the very poor behaviour that this PR fixes).
